deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/Convolution3DLayer.java

```diff
@@ -0,0 +1,257 @@
+        } else {
+            pad = layerConfig.getPadding();
+            outSize = Convolution3DUtils.get3DOutputSize(
+                    input, kernel, strides, pad, convolutionMode, dilation, isNCDHW);
+        }
+        int outD = outSize[0];
+        int outH = outSize[1];
+        int outW = outSize[2];
+
+        INDArray weightGradView = gradientViews.get(Convolution3DParamInitializer.WEIGHT_KEY);
+
+        INDArray outEpsilon = Nd4j.create(miniBatch * outChannels * outH * outW * outD);
+        INDArray reshapedEpsilon;
+        if (isNCDHW)
+            reshapedEpsilon = outEpsilon.reshape('c', miniBatch, outChannels, outD, outH, outW);
+        else
+            reshapedEpsilon = outEpsilon.reshape('c', miniBatch, outD, outH, outW, outChannels);
+
+
+        int[] intArgs = new int[]{
+                kernel[0], kernel[1], kernel[2],
+                strides[0], strides[1], strides[2],
+                pad[0], pad[1], pad[2],
+                dilation[0], dilation[1], dilation[2],
+                convolutionMode == ConvolutionMode.Same ? 0 : 1,
+                isNCDHW ? 1 : 0
+        };
+
+        INDArray delta;
+        IActivation afn = layerConfig.getActivationFn();
+        Pair<INDArray, INDArray> p = preOutput(true, true);
+        delta = afn.backprop(p.getFirst(), epsilon).getFirst();
+
+        INDArray bias;
+        INDArray biasGradView = null;
+
+        CustomOp op;
+        if (layerConfig.hasBias()) {
+
+            biasGradView = gradientViews.get(Convolution3DParamInitializer.BIAS_KEY);
+            bias = getParamWithNoise(Convolution3DParamInitializer.BIAS_KEY, true);
+
+            op = DynamicCustomOp.builder("conv3dnew_bp")
+                    .addInputs(input, weights, bias, delta)
+                    .addIntegerArguments(intArgs)
+                    .addOutputs(reshapedEpsilon, weightGradView, biasGradView)
+                    .callInplace(false)
+                    .build();
+        } else {
+            op = DynamicCustomOp.builder("conv3dnew_bp")
+                    .addInputs(input, weights, delta)
+                    .addIntegerArguments(intArgs)
+                    .addOutputs(reshapedEpsilon, weightGradView)
+                    .callInplace(false)
+                    .build();
+        }
+        Nd4j.getExecutioner().exec(op);
+
+        Gradient retGradient = new DefaultGradient();
+        if (layerConfig.hasBias()) {
+            retGradient.setGradientFor(Convolution3DParamInitializer.BIAS_KEY, biasGradView);
+        }
+        retGradient.setGradientFor(Convolution3DParamInitializer.WEIGHT_KEY, weightGradView, 'c');
+        weightNoiseParams.clear();
+
+        return new Pair<>(retGradient, reshapedEpsilon);
+    }
+
+
+    @Override
+    public INDArray preOutput(boolean training) {
+        return preOutput(training, false).getFirst();
+    }
+
+    protected Pair<INDArray, INDArray> preOutput(boolean training, boolean forBackprop) {
```

Comment:
There should be an applyDropoutIfNecessary call here somewhere?

Reply:
that's taken care of in the `activate` method in the super class.